font/src/ft/mod.rs

```diff
@@ -529,6 +589,86 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Don't try to upscale
+    if fixup_factor > 1.0 {
+        return bitmap_glyph;
+    }
+
+    let bitmap_width = bitmap_glyph.width as f64;
+    let bitmap_height = bitmap_glyph.height as f64;
+
+    let width = (bitmap_width * fixup_factor) as usize;
+    let height = (bitmap_height * fixup_factor) as usize;
+
+    let bitmap_width = bitmap_width as usize;
+    let bitmap_height = bitmap_height as usize;
+
+    let b_buf = &bitmap_glyph.buf;
+    let scaling_factor =
+        (bitmap_width as f32 / width as f32).max(bitmap_height as f32 / height as f32);
+    let advance_step = scaling_factor.ceil() as usize;
+    let mut scaled_buffer = Vec::with_capacity(width * height * 3);
+
+    let mut new_line_index = 0;
+    let mut source_line_index = 0;
+
+    while new_line_index < height {
+        let mut new_column_index = 0;
+        let mut source_column_index = 0;
+
+        while new_column_index < width {
+            let mut r: u32 = 0;
+            let mut g: u32 = 0;
+            let mut b: u32 = 0;
+            let mut pixels_picked: u32 = 0;
+
+            let source_end_line = std::cmp::min(source_line_index + advance_step, bitmap_height);
+            let source_end_column = std::cmp::min(source_column_index + advance_step, bitmap_width);
+
+            let mut source_line_index = source_line_index;
+            while source_line_index < source_end_line {
+                let cur_pixel_index = source_line_index * bitmap_width * 3;
+
+                let mut source_column_index = source_column_index;
+                while source_column_index < source_end_column {
+                    r += b_buf[cur_pixel_index + source_column_index * 3] as u32;
+                    g += b_buf[cur_pixel_index + source_column_index * 3 + 1] as u32;
+                    b += b_buf[cur_pixel_index + source_column_index * 3 + 2] as u32;
+                    source_column_index += 1;
+                    pixels_picked += 1;
+                }
+                source_line_index += 1;
+            }
+
+            if pixels_picked == 0 {
+                scaled_buffer.push(0);
+                scaled_buffer.push(0);
+                scaled_buffer.push(0);
+            } else {
+                scaled_buffer.push((r / pixels_picked) as u8);
+                scaled_buffer.push((g / pixels_picked) as u8);
+                scaled_buffer.push((b / pixels_picked) as u8);
+            }
+
+            source_column_index += advance_step;
+            new_column_index += 1;
+        }
+        source_line_index += advance_step;
+        new_line_index += 1;
+    }
+
+    // This top computation performs better with a scaling algo we use.
+    bitmap_glyph.top = ((bitmap_glyph.top as f32 * fixup_factor as f32) as i32
+        + bitmap_glyph.top / advance_step as i32)
+        / 2;
```

Comment:
Where exactly does this formula come from? Is this just using metrics based on what's reasonable, or is there any source recommending this strategy?

Reply:
This formula came from hours of testing and comparing with different font sizes etc. You can try to compare it to just fixup * top.

Reply:
I'm not fond of the idea of just testing out random combinations of font metrics until something seems to work out okay, this usually means that the result is over-specialized for the fonts installed when testing, rather than being a good general purpose solution.

We should try to see either what other implementations do, or if there's any reference on how these metrics should be scaled. Otherwise we need to put some reason behind the computation, that is more than just 'it seems to look okay this way'.

Reply:
FWIW, I simply scale it and `ceil` it:
```c
ceil(font->face->glyph->bitmap_top * font->pixel_size_fixup)
```
This scaled top value is then used as offset from the primary font's baseline.

It looks reasonable to me, with the (granted, few) font combinations I've tried: ![grim-snapshot-cca14db4-469d-47d3-952b-c4f09a52a599](https://user-images.githubusercontent.com/14031526/69906853-3e6eed00-13ca-11ea-996e-ed6a1ff0a953.png)

(_Fantasque Sans Mono_ + _JoyPixels_)